src/vptranslator.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+impl PartialEq for ValidityPeriod {
+    fn eq(&self, other: &Self) -> bool {
+        (self.start_date, &self.end_date) == (other.start_date, &other.end_date)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum ExceptionType {
+    Sub,
+    Add,
+}
+
+#[derive(Debug)]
+struct ExceptionDate {
+    date: NaiveDate,
+    exception_type: ExceptionType,
+}
+
+impl PartialEq for ExceptionDate {
+    fn eq(&self, other: &ExceptionDate) -> bool {
+        self.exception_type == other.exception_type
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct BlockPattern {
+    week: u8,
+    nb_weeks: i64,
+    operating_days: Vec<Weekday>,
+    validity_periods: Vec<ValidityPeriod>,
+    exceptions: Vec<ExceptionDate>,
+}
+
+fn get_prev_monday(date: NaiveDate) -> NaiveDate {
+    let res = date + Duration::days(-1 * date.weekday().num_days_from_monday() as i64);
+    res
+}
+
+fn weeks(dates: &BTreeSet<NaiveDate>) -> Vec<u8> {
+    let start_date: NaiveDate = get_prev_monday(*dates.iter().next().unwrap());
+    let end_date: NaiveDate = *dates.iter().next_back().unwrap();
+    let length = (end_date.signed_duration_since(start_date).num_weeks() + 1) as usize;
+
+    let mut res = vec![0; length];
+    for date in dates {
+        let w = date.signed_duration_since(start_date).num_weeks() as usize;
+        res[w] |= 1 << (7 - date.weekday().number_from_monday());
+    }
+    res
+}
+
+fn dists(w: u8, weeks: &Vec<u8>) -> u32 {
+    let mut res: u32 = 0;
+    for n in weeks.iter() {
+        if *n != 0 {
+            res += (*n as u8 ^ w).count_ones();
+        }
+    }
+    res
+}
+
+fn get_min_week_pattern(weeks: &Vec<u8>) -> u8 {
+    let mut best: u8 = 0;
+    let mut best_score = std::u32::MAX;
+    for i in 0..128 {
+        let score = dists(i, weeks);
+        if (score < best_score)
+            || ((score == best_score) && (score.count_ones() < best_score.count_ones()))
+        {
+            best_score = score;
+            best = i;
+        }
+    }
+    best
+}
+
+fn fill_exceptions(
+    start_date: NaiveDate,
+    exception: u8,
+    exception_type: ExceptionType,
+    exception_list: &mut Vec<ExceptionDate>,
+) {
+    for i in 0..7 {
+        if exception & (1 << i) == (1 << i) {
+            let date = start_date + Duration::days((6 - i) as i64);
+            exception_list.push(ExceptionDate {
+                date: date,
+                exception_type: exception_type.clone(),
+            });
+        }
+    }
+}
+
+fn fill_operating_days(week: u8, operating_days: &mut Vec<Weekday>) {
+    for i in 0..7 {
+        if week & (1 << i) == (1 << i) {
+            operating_days.push(Weekday::from_u8(6 - i).unwrap());
+        }
+    }
+    operating_days.sort_by_key(|w| w.num_days_from_monday());
+}
+
+fn clean_extra_dates(start_date: NaiveDate, end_date: NaiveDate, dates: &mut Vec<ExceptionDate>) {
+    dates.retain(|d| d.date >= start_date && d.date <= end_date);
+}
+
+pub fn translate(dates: &BTreeSet<NaiveDate>) -> BlockPattern {
```

Comment:
travis will complain, as it is required to have some documentation on every public method/struct, could you add some on this ?